modopt/opt/algorithms/gradient_descent.py

```diff
@@ -0,0 +1,354 @@
+    """
+
+    def __init__(
+        self,
+        x,
+        grad,
+        prox,
+        cost,
+        eta=1.0,
+        eta_update=None,
+        epsilon=1e-6,
+        epoch_size=1,
+        metric_call_period=5,
+        metrics=None,
+        **kwargs,
+    ):
+        # Set the initial variable values
+        if metrics is None:
+            metrics = {}
+        # Set default algorithm properties
+        super().__init__(
+            metric_call_period=metric_call_period,
+            metrics=metrics,
+            **kwargs,
+        )
+        self.iter = 0
+        self._check_input_data(x)
+        self._x_old = np.copy(x)
+        self._x_new = np.copy(x)
+        self._speed_grad = np.zeros(x.shape, dtype=float)
+        self._dir_grad = np.zeros_like(x)
+        # Set the algorithm operators
+        for operator in (grad, prox, cost):
+            self._check_operator(operator)
+        self._grad = grad
+        self._prox = prox
+        if cost == 'auto':
+            self._cost_func = costObj([self._grad, self._prox])
+        else:
+            self._cost_func = cost
+        # Set the algorithm parameters
+        for param_val in (eta, epsilon):
+            self._check_param(param_val)
+        self._eta = eta
+        self._eps = epsilon
+
+        # Set the algorithm parameter update methods
+        self._check_param_update(eta_update)
+        self._eta_update = eta_update
+        self.idx = 0
+        self.epoch_size = epoch_size
+
+
+    def iterate(self, max_iter=150):
+        """Iterate.
+
+        This method calls update until either convergence criteria is met or
+        the maximum number of iterations is reached.
+
+        Parameters
+        ----------
+        max_iter : int, optional
+            Maximum number of iterations (default is ``150``)
+
+        """
+        self._run_alg(max_iter)
+
+        # retrieve metrics results
+        self.retrieve_outputs()
+
+        self.x_final = self._x_new
+
+    def _update(self):
+        self._grad.get_grad(self._x_old)
+        self.update_grad_dir(self._grad.grad)
+        self.update_grad_speed(self._grad.grad)
+        step = self._eta / (np.sqrt(self._speed_grad) + self._eps)
+        self._x_new = self._x_old - step * self._dir_grad
+        if self.idx % self.epoch_size == 0:
+            self.reset()
+            self.update_reg(step)
+        self._x_old = self._x_new.copy()
+        if self._eta_update is not None:
+            self._eta = self._eta_update(self._eta, self.idx)
+        # Test cost function for convergence.
+        if self._cost_func:
+            self.converge = (
+                self.any_convergence_flag()
+                or self._cost_func.get_cost(self._x_new)
+            )
+
+    def update_grad_dir(self, grad):
+        """Update the gradient descent direction."""
+        self._dir_grad = grad
+
+    def update_grad_speed(self, grad):
+        """Update the gradient descent speed."""
+        pass
```

Comment:
I guess here you are simply defining a method you expect the child classes to override, right? I don't really think this is needed unless the child class is forced to override this (*e.g.* as an abstract method).

Reply:
I had to defined this empty method for the generic case. (cf `_update` method, which is commont to all gradient descent algorithms.